Fix to #7115: Allow setting "hardFactor"

Volfee opened the pull request:
## Pull Request template

## Purpose / Description
Hard Factor is an option of V2 Scheduler, but it is not possible to modify it from Deck Options.

## Fixes
Fixes #7115

## Approach
* Modified DeckOptions to include hardFactor and added selection field to Deck Options / Reviews tab.
* Hard Factor can be modified irrespective of user using SchedV1 / SchedV2, but this option is set in general preferences not in deck options. I'm not sure how to make it disabled if SchedV1 is used.

## How Has This Been Tested?
* Tested on emulated device - Pixel 3A. SDK 29.
* Added unit test that tests if value is not set for newly created deck and returns correct value after value is modified.

## Learning (optional, can help others)
Hard Factor is described here: https://anki.tenderapp.com/discussions/ankidesktop/31735-what-does-hard-interval-mean

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)

_edited by @david-allison-1_

david-allison-1 reviewed (changes requested):
Looks good!

The main check would be to ensure that we're not applying or displaying this property for SchedV1.

`mCol.schedVer()` can be used to get the version, extract `isSchedV2()` into a method, and use this to optionally set both the value in the values set, and remove the preference from visibility.

 I don't exactly know how to remove a preference, but to get started:

https://github.com/ankidroid/Anki-Android/blob/517b5a15ea15761fc0a816cb35daa53ec3910438/AnkiDroid/src/main/java/com/ichi2/anki/DeckOptions.java#L674

is where the preferences are added.

I'd recommend looking at `this.getPreferenceScreen()`

You can generally get a preference by the key, and remove it like that.

If you're having problems, give me a shout

---

A second note is that we're a rebase-heavy repository and avoid commits. It's best to rebase your current commits onto master and then `git push --force-with-lease` to the PR  (Android Studio's default for force pushing) to get a clean history.

mikehardy commented:
@Volfee marked this as waiting for feedback for unresolved comment

Volfee commented:
Hey, sorry for not following up yet! I managed to figure out how to make option disabled when SchedV2 is not used (thanks David for advice). Will send improved PR over the weekend.

mikehardy commented:
Excellent! No specific rush, I mostly commented with a tag just to make sure you'd seen it - some people have notifications set such that they only see direct tags :-). Cheers

Volfee commented:
Hey, I'm a bit confused by git workflow -> I implemented changes on hardFactor branch and rebased it to the master on my forked repository following Davids advice (https://github.com/Volfee/Anki-Android/commit/815cc5fd24f1959565869f64e8d07d7df59ab899). 

Now, should I create a separate pull request to add changes or is there a way to include those changes to this pull request? I suspect that both can be done, but what would be the better way in this case?

Thanks!

mikehardy commented:
I always do it by the command line, so forgive me if this is contradictory (the result should be the same though?)

- assume you have a git remote named "upstream"
- I do a git co master then a git pull upstream/master to get all the upstream changes on my local device
- I git co branchname and git rebase origin master so I have those on my branch now (resolve conflicts if needed)
- I git push origin +branchname to clobber github state with my newly-master-integrated changes

that's it?

Volfee commented:
Thanks Mike, I faced some troubles, but I think I managed to do it. Your comment was helpful.

david-allison-1 commented on AnkiDroid/src/main/res/xml/deck_options.xml:
```diff
@@ -37,11 +37,11 @@
                 android:negativeButtonText="@string/dialog_cancel"
                 android:positiveButtonText="@string/dialog_positive_remove"
-                android:title="@string/dialog_positive_remove" />
+                android:title="@string/deck_conf_remove" />
```
```suggestion
                android:title="@string/dialog_positive_remove" />
```

david-allison-1 commented on AnkiDroid/src/main/res/xml/deck_options.xml:
```diff
@@ -37,11 +37,11 @@
                 android:dialogIcon="@drawable/ic_dialog_alert"
                 android:dialogMessage="@string/deck_conf_remove_message"
-                android:dialogTitle="@string/dialog_positive_remove"
+                android:dialogTitle="@string/deck_conf_remove_title"
```
```suggestion
                android:dialogTitle="@string/dialog_positive_remove"
```

david-allison-1 reviewed (changes requested):
Took on a conflict from b82ccc4331848a28e93f6d14feb2934ca7fa31f8 and bab55f316f90de917874c9fac7d4902998dbad69

We've been doing a lot of work deduplicating translations 

(#7417)

david-allison-1 reviewed (approved):
Awaiting #7467

mikehardy reviewed (approved):
This looks good to me and all the issues with CI are external to the PR - I pulled this branch, rebased it to the recently merged test fixes and hard-pushed it back to your branch @Volfee - it should pass CI now and I'll merge it, thanks!

mikehardy commented:
boom! merged :-)
